### ISSUE TITLE
Rule 901162 in v4.0.0 for tx.allowed_request_content_type is now blocking CSP violation reports, type application/csp-report. These were explicitly allowed in v3.2.2

### ISSUE THREAD

**Issue by Simon1954** [coreruleset/coreruleset#3138]
### Your Environment

* CRS version (e.g., v3.2.0):
* Paranoia level setting:
* ModSecurity version (e.g., 2.9.3):
* Web Server and version (e.g., apache 2.4.41):
* Operating System and version:

### Confirmation

[ ] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by Simon1954** [comment]
CRS v4.0.0-rc1
Apache v2.4.55 on Linux Ubuntu 18.04.6 LTS with PHP 8.2.3
This CRS version is preventing any Content-Security-Policy violation reports from being processed because 
application/csp-report is no longer included by default in rule 901162 which sets tx.allowed_request_content_type. It was included in CRS v 3.2.0, which allowed CSP reports to be processed.
Tested using latest versions of Firefox, Chrome and Edge.

**Comment by RedXanadu** [comment]
Hi @Simon1954.

Did you close this issue because you already found the answer you were looking for?

In any event, `application/csp-report` was intentionally removed from the default list of allowed content types a few months ago, in this PR: #2768.

Any content types that do not have an associated default body parser (allowing for the request payload to be properly inspected) were removed. Otherwise, it may have been possible to bypass detection by exploiting a content type that did not have a proper body parser assigned to allow for inspection.

If you're working in an environment where `application/csp-report` is used and you can say for certain that it will take the form of JSON requests, for example, then you can add that content type to your allowed list. Just make sure to enable JSON body processing for requests of type `application/csp-report`, and you're good to go!

**Comment by Simon1954** [comment]
Hi Andrew,

 

Thanks for the quick reply. I didn’t intentionally close my query – just a late night mistake.

 

I was under the impression that all modern browsers which implement the CSP protocol will send reports (if specified in the CSP header) with data posted in a JSON-encoded format and will use the application/csp-report content type. The CSP protocol requires the user’s browser to post such reports to a URI (or maybe a list of them) specified in the CSP header. In my case at least, that URL will actively check for a suitable JSON-encoded set of data about the violation, and then update server logs accordingly and return a http 204 code. (If the JSON decoding fails then it is just ignored.) Since websites are now actively recommended for security reasons to use CSP headers (see for example Immuniweb, which downgrades the security of websites which don’t), it seems strange that this CSP reporting content-type should be blocked in your default core ruleset. 

 

On some but not all of the websites I manage, I started to become suspicious at the absence of any CSP violation reports on some of them (which had been quite common a while ago), while others with identical scripts and Apache server config parameters were logging them. It took quite a lot of time to track down the cause before I realised that it was the CRS 4.0.0 version of the core ruleset installed on some of the websites vs 3.2.2 on others which was explanation. I suspect others using CSP headers and reports will run into the same problem.

 

Since the 4.0.0 core ruleset does allow the application/json content type, I don’t see why allowing the application/csp-report content type introduces any additional security risk, since genuine data in the latter is in JSON format anyway. If a malicious person wanted to misuse this content type they could just as easily send bad content labelled as application/json.

 

Best wishes
Simon Hathrell

**Comment by dune73** [comment]
Thank you for your extensive reasoning @Simon1954 and thank you for highlighting this problem.

The situation is: ModSecurity will not parse `application/csp-report` by default. CRS is therefore blind to almost all attacks in the CSP report. As @RedXanadu explained above, we have decided to deny requests that are not parsed by ModSecurity so you do not get a false sense of security (the situation with CRS 3.2.2 is that).

Now given the fact that the CSP reports are in fact JSON, ModSecurity **could** parse it, if instructed properly. The instruction to do so currently has to be placed in the so-called ModSecurity recommended rules. These rules are distributed by ModSecurity and not by CRS.

We have decided to replace the ModSec recommended rules with our own set of (recommended) rules after the release of CRS v4. Namely, because this would give us flexibility and it would allow us to fix quite a few rough edges. Your report and the whole CSP report situation that you brought to our attention falls into this category as well. I think it's something we need to address in our own set of rules.

As for an immediate workaround: Reconfigure the recommended rules as Xanadu instructed you and add the CT type to the CRS list of allowed types.
